python/pfs/drp/stella/lsf.py

```diff
@@ -0,0 +1,688 @@
+            left = self.center
+            right = self.length - self.center - 1
+            middle = max(left, right)
+            length = 2*middle + 1
+            kernel = np.zeros(length, dtype=self.values.dtype)
+            kernel[middle - left:middle + right + 1] = self.values
+        else:
+            kernel = self.values
+        return np.convolve(array, kernel, "same")
+
+    def toArray(self, length, center, interpolator=DEFAULT_INTERPOLATOR):
+        """Convert to an array at a nominated position
+
+        Parameters
+        ----------
+        length : `int`
+            Desired length of array.
+        center : `float`
+            Index position at which to insert the kernel.
+        interpolator : callable, optional
+            Interpolator factory. Calling this with arrays of ``x`` and ``y``
+            should produce an interpolator that can be called with an array of
+            ``x`` values at which to interpolate.
+
+        Returns
+        -------
+        array : `numpy.ndarray`, floating-point
+            Array containing inserted kernel.
+        """
+        return interpolator(self.indices + center, self.values)(np.arange(length))
+
+    def computeStdev(self):
+        """Compute standard deviation of kernel"""
+        centroid = np.sum((self.indices*self.values).astype(np.float64))
+        return np.sqrt(np.sum((self.values*(self.indices - centroid)**2).astype(np.float64)))
+
+    def interpolate(self, indices):
+        """Interpolate the kernel at provided indices
+
+        Parameters
+        ----------
+        indices : array_like, floating-point
+            Indices at which to interpolate.
+
+        Returns
+        -------
+        values : array_like, floating-point
+            Values of kernel at the provided indices.
+        """
+        return np.interp(indices, self.indices, self.values, left=0.0, right=0.0)
+
+
+class Lsf(ABC):
+    """Abstract base class for a Line-Spread Function
+
+    A line-spread function (LSF) is the one-dimensional version of a
+    point-spread function (PSF). It models the response of the spectrograph to a
+    single infinitely thin line, which may be a function of position.
+
+    This class knows nothing about wavelengths: all positions are specified in
+    pixels.
+
+    Subclasses should implement the abstract ``computeKernel`` method; other
+    methods have default implementations that rely on that.
+
+    This class includes two main styles of methods:
+
+    - methods taking a ``center`` position of type `float`. These methods are
+      intended for interacting with a numpy array.
+    - methods taking a ``point`` of type `lsst.geom.Point2D` (optionally; if not
+      provided, the ``point`` defaults to the result of ``getAveragePosition``).
+      These methods are  intended for emulating (duck-typing) an
+      `lsst.afw.detection.Psf` for when interacting with a
+      `lsst.afw.image.Image`.
```

Comment:
Please provide an example (e.g. subtracting a spectrum specified as a set of (intensity, lambda) values)